### DISCUSSION TITLE
generator에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by hwangstar156** [woowacourse-study/2022-woowahan-bansanghwe#110] [JavaScript]
## generator

제너레이터란 코드블록의 실행을 일시 중지했다가 필요한 시점에 재개할 수 있는 특수한 함수입니다.

그렇다면 제너레이터 함수와 일반 함수의 차이는 무엇일까요??

- 제너레이터 함수는 함수 호출자에게 함수 실행의 제어권을 양도할 수 있다.
- 제너레이터 함수는 함수 호출자와 함수의 상태를 주고 받을 수 있다.
- 제너레이터 함수를 호출하면 제너레이터 객체를 반환한다. 

## 제너레이터 객체

제너레이터 함수를 호출하면 함수 코드 블록을 실행하는 것이 아닌 제너레이터 객체를 생성하고 반환한다. (이때 제너레이터 객체는 **이터러블**임과 동시에 **이터레이터**이다.)

## 제너레이터의 일시 중지와 재개

제너레이터는 yield와 next 메서드를 통해 실행을 일시 중지했다가 필요한 시점에 다시 재개할 수 있다.
yield의 뜻은 `양도`로 원래 일반 함수는 제어권은 모두 함수가 가지고 있지만 yield 메서드를 통해 호출자에게 제어권을 양도한다는 의미를 가진다.

제너레이터 함수는 yield 표현식을 만날때까지 함수를 실행하고 yield를 만나면 함수를 멈추고 next를 호출하면 다시 다음 yield를 만날때까지 진행한다.

그리고 이 yield 키워드는 멈춘다라는 기능외에도 yield뒤에 있는 표현식의 평가 결과를 호출자에게 반환한다.

또 next 키워드는 `generator.next()` 이런 식으로 호출을 하는데 next내부에 인자를 넣어서 yield표현식에게 값을 전달할 수 있다.

## 제너레이터의 활용

그렇다면 이 제너레이터는 어디에 사용할 수 있을까?

현재 비동기 처리에서 사용되는 async/await에 사용할 수 있고 redux-saga에서는 제너레이터 문법을 이용하여 비동기를 처리하고 있다.

**Comment by airman5573** [comment]
궁금하다...자수밍의 제너레이터

**Comment by kamwoo** [comment]
자바스크립트의 제너레이터에 대해서 설명해주세요

**Comment by kamwoo** [comment]
제너레이터 객체에 대해서 설명해주세요

**Comment by kamwoo** [comment]
이터러블과 이터레이트 객체는 뭘까요?

**Comment by kamwoo** [comment]
제너레이터가 아닌 이터레이터 객체는 무엇이 있나요?

**Comment by kamwoo** [comment]
제너레이터는 언제 많이 쓰나요?

**Comment by kamwoo** [comment]
async-await에서는 어떻게 사용하나요?

**Comment by kamwoo** [comment]
reduc-saga에서는 어떻게 사용되나요?

**Comment by kamwoo** [comment]
redux-saga는 왜 제어레이트를 사용했을까요?

**Comment by kamwoo** [comment]
제너레이터의 단점은 무엇이 있을까요?

**Comment by kamwoo** [comment]
redux-thunk 대신에 saga를 이용하실건가요?

**Comment by kamwoo** [comment]
helper함수들에 직접적인 도움을 받은 경험이 있나요?

**Comment by kamwoo** [comment]
디바운싱은 무엇인가요?

**Comment by kamwoo** [comment]
쓰로틀링은 뭔가요?

**Comment by kamwoo** [comment]
콜백 헬을 경험해보신 적이 있나요?

**Comment by kamwoo** [comment]
Promise의 장점은 무엇이 있을까요?

**Comment by kamwoo** [comment]
Promise에 대해서 설명해주실 수 있나요?

**Comment by kamwoo** [comment]
fetch 동작 방식은 뭔가요

**Comment by kamwoo** [comment]
제너레이터를 실행 컨텍스트 관점에서 설명해주실 수 있나요?

**Comment by kamwoo** [comment]
제너레이터를 왜 많이 사용하지 않을까요

**Comment by kamwoo** [comment]
순열과 조합에 대해서 설명해주실 수 있나요

**Comment by kamwoo** [comment]
자바스크립트에서 비동기란 무엇일까요
